pkg/processor/schemaregistry/avro/reflect.go

```diff
@@ -0,0 +1,254 @@
+// Copyright © 2023 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package avro
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
+	"github.com/conduitio/conduit/pkg/record"
+	"github.com/hamba/avro/v2"
+)
+
+var (
+	structuredDataType = reflect.TypeOf(record.StructuredData{})
+	byteType           = reflect.TypeOf(byte(0))
+)
+
+//nolint:gocyclo // reflection requires a huge switch, it's fine
+func reflectInternal(path []string, v reflect.Value, t reflect.Type) (avro.Schema, error) {
+	if t == nil {
+		return nil, cerrors.New("can't get schema for untyped nil") // untyped nil
+	}
+	switch t.Kind() {
+	case reflect.Bool:
+		return avro.NewPrimitiveSchema(avro.Boolean, nil), nil
+	case reflect.Int64, reflect.Uint32:
+		return avro.NewPrimitiveSchema(avro.Long, nil), nil
+	case reflect.Int, reflect.Int32, reflect.Int16, reflect.Uint16, reflect.Int8, reflect.Uint8:
+		return avro.NewPrimitiveSchema(avro.Int, nil), nil
+	case reflect.Float32:
+		return avro.NewPrimitiveSchema(avro.Float, nil), nil
+	case reflect.Float64:
+		return avro.NewPrimitiveSchema(avro.Double, nil), nil
+	case reflect.String:
+		return avro.NewPrimitiveSchema(avro.String, nil), nil
+	case reflect.Pointer:
+		var vElem reflect.Value
+		if v.IsValid() {
+			vElem = v.Elem()
+		}
+		s, err := reflectInternal(path, vElem, t.Elem())
+		if err != nil {
+			return nil, err
+		}
+
+		var schemas avro.Schemas
+		if us, ok := s.(*avro.UnionSchema); ok && us.Nullable() {
+			// it's already a nullable schema
+			return s, nil
+		} else if ok {
+			// take types from union schema
+			schemas = us.Types()
+		} else if s.Type() != avro.Null {
+			// non-nil type
+			schemas = avro.Schemas{s}
+		}
+
+		s, err = avro.NewUnionSchema(append(schemas, &avro.NullSchema{}))
+		if err != nil {
+			return nil, err
+		}
+
+		return s, nil
+	case reflect.Interface:
+		if !v.IsValid() || v.IsNil() {
+			// unknown type, fall back to nullable string
```

Comment:
If we didn't do that, a field could be lost, it would be like it's not even there. For example, when we get a value like `map[string]any{"foo":nil}` we have no idea what `nil` could be, so we fall back to the nullable string.

I'll mention it in code 👍